### PR TITLE
Change folder structure for jbuilder

### PR DESCRIPTION
**Pull request by rgrinberg**
The new project structure is such that every package is self contained in its
own directory.

I've also added the cohttp-top package (which was missing in the original port)

### CODE REVIEW

**Comment by rgrinberg** [comment]
I've also removed all the manual tests. Nobody runs these manually so they're really not helping much. If someone would like to bring them back, please turn them into a proper test suite first.

**Comment by rgrinberg** [comment]
This revision also reintroduces the various async/lwt curl/static serve binaries. I've elected to put them in their own packages because they have their own dependencies (like cmdliner) for example.

File: cohttp-lwt-unix-bin/jbuild [hunk 1]
```diff
@@ -0,0 +1,7 @@
+(jbuild_version 1)
+
+(executables
+ ((names (cohttp_curl_lwt cohttp_proxy_lwt cohttp_server_lwt))
+  (libraries (cohttp-lwt-unix cohttp_server cmdliner))
+  (package cohttp-lwt-unix-bin)
```
Comment (avsm) [inline cohttp-lwt-unix-bin/jbuild]: Is it really worth having a separate OPAM package for the binaries? The cmdliner dependency is pretty minor and doesnt have a runtime impact on the cohttp_lwt_unix library. I'd suggest folding cohttp-lwt-unix-bin back into cohttp-lwt-unix

File: cohttp-top.opam [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+tags: ["org:mirage" "org:xapi-project"]
+dev-repo: "https://github.com/mirage/ocaml-cohttp.git"
+build: [
+  ["jbuilder" "subst" "-n" name] {pinned}
+  ["jbuilder" "build" "-p" name "-j" jobs]
+]
+build-test: [["jbuilder" "runtest" "-p" name "-j" jobs]]
+depends: [
+  "jbuilder" {build & >= "1.0+beta10"}
+  "cohttp"
```
Comment (avsm) [inline cohttp-top.opam]: might as well put a minimum version in here that is greater than the current released cohttp, to ensure that --dev picks up a useful version

File: cohttp/src/jbuild [hunk 3]
```diff
@@ -2,7 +2,7 @@
 #require "unix"
 
 let version =
-  let ic = open_in "../VERSION" in
+  let ic = open_in "../../VERSION" in
```
Comment (avsm) [inline cohttp/src/jbuild]: there's got to be a better way to do version stamping, but not a blocker to merge :-)

Comment (rgrinberg) [inline cohttp/src/jbuild]: This is the workaround I got from jsoo ^_^

**Comment by rgrinberg** [comment]
@avsm Halp. looks like the version constraints are making the CI fail.

File: cohttp/src/jbuild (hunk 3 above)
Comment (avsm) [inline cohttp/src/jbuild]: @hhugo @diml -- is it worth an upstream jbuilder issue to track how to improve creating version file stamps?  probably good not to to propagate too many hacky rules throughout the ecosystem.

File: cohttp-async/bin/jbuild [hunk 4]
```diff
@@ -0,0 +1,7 @@
+(jbuild_version 1)
+
+(executables
+ ((names (cohttp_curl_async cohttp_server_async))
+  (package cohttp-async-bin)
```
Comment (avsm) [inline cohttp-async/bin/jbuild]: I think this needs to be cohttp-async, which is killing the build (not constraints).  Same with the cohttp-lwt-unix jbuild

File: cohttp/src/jbuild (hunk 3 above)
Comment (samoht) [inline cohttp/src/jbuild]: I though it was why we had `%%VERSION%%` and `opam subst` :-)

**Comment by rgrinberg** [comment]
@avsm that was indeed a bug, but it wasn't the cause of the CI failures

**Comment by avsm** [comment]
Oh I see...

```
The following dependencies couldn't be met:
  - cohttp-async -> cohttp > 0.22.0
Your request can't be satisfied:
  - cohttp>0.22.0 is not available because the package is pinned to /repo, version 0.22.0.
```

**Comment by rgrinberg** [comment]
I'll revert those constraints for now

File: cohttp/src/jbuild (hunk 3 above)
Comment (diml) [inline cohttp/src/jbuild]: Now you can also do:

```scheme
(rule (with-stdout-to version.ml "let version = {|${version:foo}|}\n"))
```

**Pull request merged**